File: python/run-length-encoding/run_length_encoding.py
def encode(string):
    def to_append():
        num = str(count) if count > 1 else ""
        return "" if prev_char is None else num + prev_char

    if not string:
        return ""

    prev_char = None
    count = 1
    out = ""
    for char in string:
        if char == prev_char:
            count += 1
        else:
            out += to_append()
            prev_char = char
            count = 1
    return out + to_append()


def decode(string):
    number = ""
    out = ""
    for char in string:
        if char.isdigit():
            number += char
        else:
            repeats = int(number) if number else 1
            out += repeats * char
            number = ""
    return out

File: csharp/run-length-encoding/RunLengthEncoding.cs
using System;

public static class RunLengthEncoding
{
    public static string Encode(string input)
    {
        if (input.Length == 0) return "";
        char? prevChar = null;
        int count = 1;
        string output = "";
        foreach(char ch in input)
        {
            if (ch == prevChar)
                count++;
            else
            {
                output += ToAppend(count, prevChar);
                prevChar = ch;
                count = 1;
            }
        }
        return output + ToAppend(count, prevChar);
    }

    public static string Decode(string input)
    {
        var number = "";
        var output = "";
        foreach (char ch in input)
        {
            if (Char.IsDigit(ch))
                number += ch;
            else
            {
                var repeats = number.Length > 0 ? int.Parse(number) : 1;
                output += new string(ch, repeats);
                number = "";
            }
        }
        return output;
    }

    private static string ToAppend(int count, char? prevChar)
    {
        var num = count > 1 ? count.ToString() : "";
        return prevChar.HasValue ? num + prevChar : "";
    }
}
